elastic/kibana
# [Dashboard] Fix Alias Redirect Race Condition

## Summary

This PR fixes a race condition that could happen in the case of an alias match redirect. Before this PR it was possible for a URL update (redirecting to the alias) to happen in the middle of building the errored state of the Dashboard. This would sometimes result in the error from the last dashboard (the old saved object id) being surfaced instead of the redirect properly firing.

To fix this, this PR allows the Dashboard factory to return undefined in the case of a redirect and updates the renderer to support that.

### Before
https://github.com/elastic/kibana/assets/14276393/fa7ef107-3bff-43c1-9174-482e91179cbb

### After
![After](https://github.com/elastic/kibana/assets/14276393/d265534f-0a3f-4e7c-8b32-a23235c74342)



### A note on tests
It seems like this alias match functionality breaks often enough that we should cover it with functional tests. Testing this will be difficult because we'll need to spoof a situation where a redirect would be required - which requires multiple spaces, and saved objects from 7.17. This test coverage should be added as a followup because it is important that this PR merged into 8.9.

### How to test manually?
1. Extract this file (containing ES Data with 7.17 ID conflicts).
[esData backup (7.17 ID conflicts).zip](https://github.com/elastic/kibana/files/11923254/esData.backup.7.17.ID.conflicts.zip)
2. Start ES pointing to the extracted folder. ` yarn es snapshot -E path.data="/path/to/data"`
3. Load into the `Super Cool New Space`
4. Paste in URL with an old ID to that Dashboard `{Your address}/s/super-cool-new-space/app/dashboards#/view/326754f0-0ac2-11ee-9f77-63b7bd4c6a36`
5. The Dashboard should redirect properly to the new Dashboard with the ID `ce23f0a9-47e2-5cf5-82b2-9446e880a0ae`
6. Paste the link from step 4 into your URL bar as many times as you want... it should keep redirecting properly!
